ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/validation/RequestValidator.java

```diff
@@ -0,0 +1,172 @@
+ */
+public class RequestValidator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RequestValidator.class);
+
+  private final Map<Class<? extends Statement>, StatementValidator> customValidators;
+  private final Function<ServiceContext, SchemaInjector> schemaInjectorFactory;
+  private final Supplier<KsqlExecutionContext> snapshotSupplier;
+  private final ServiceContext serviceContext;
+  private final KsqlConfig ksqlConfig;
+
+  /**
+   * @param customValidators        a map describing how to validate each statement of type
+   * @param schemaInjectorFactory   creates a {@link SchemaInjector} for schema validation
+   * @param snapshotSupplier        supplies a snapshot of the current execution state, the
+   *                                snapshot returned will be owned by this class and changes
+   *                                to the snapshot should not affect the source and vice versa
+   * @param serviceContext          the {@link ServiceContext} to use
+   * @param ksqlConfig              the {@link KsqlConfig} to validate against
+   */
+  public RequestValidator(
+      final Map<Class<? extends Statement>, StatementValidator> customValidators,
+      final Function<ServiceContext, SchemaInjector> schemaInjectorFactory,
+      final Supplier<KsqlExecutionContext> snapshotSupplier,
+      final ServiceContext serviceContext,
+      final KsqlConfig ksqlConfig
+  ) {
+    this.customValidators = Objects.requireNonNull(customValidators, "customValidators");
+    this.schemaInjectorFactory = Objects.requireNonNull(schemaInjectorFactory, "schemaInjector");
+    this.snapshotSupplier = Objects.requireNonNull(snapshotSupplier, "snapshotSupplier");
+    this.serviceContext = Objects.requireNonNull(serviceContext, "serviceContext");
+    this.ksqlConfig = Objects.requireNonNull(ksqlConfig, "ksqlConfig");
+  }
+
+  /**
+   * Validates the messages against a snapshot in time of the KSQL engine.
+   *
+   * @param statements          the list of statements to validate
+   * @param propertyOverrides   a map of properties to override for this validation
+   * @param sql                 the sql that generated the list of statements, used for
+   *                            generating more useful debugging information
+   *
+   * @return the number of new persistent queries that would be created by {@code statements}
+   * @throws KsqlException if any of the statements cannot be validated, or the number
+   *                       of requested statements would cause the execution context
+   *                       to exceed the number of persistent queries that it was configured
+   *                       to support
+   */
+  public int validate(
+      final List<ParsedStatement> statements,
+      final Map<String, Object> propertyOverrides,
+      final String sql
+  ) {
+    final KsqlExecutionContext ctx = snapshotSupplier.get();
+    final SchemaInjector injector = schemaInjectorFactory.apply(
```

Comment:
shouldn't be a need to `SandboxedServiceContext.create(serviceContext)` as the validator should be passed a sandboxed service context in the first place.